### PR TITLE
Gamma correction

### LINKED ISSUE #2169: Gamma correction

### PR DESCRIPTION
**Pull request by mcquin**
Resolves #2169

### CODE REVIEW

File: cellprofiler/modules/gammacorrection.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+"""
+
+Gamma correction
```
Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: Documentation?

Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: Test?

File: cellprofiler/modules/gammacorrection.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+"""
+
+import cellprofiler.image
+import cellprofiler.module
+import cellprofiler.setting
+import skimage.exposure
+
+
+class GammaCorrection(cellprofiler.module.Module):
+    category = "Image processing"
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Capitalize `"processing"` (or change all existing `"Image Processing"`).

File: cellprofiler/modules/gammacorrection.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    module_name = "Gamma correction"
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Current style is `"GammaCorrection"`.

File: cellprofiler/modules/gammacorrection.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    def display(self, workspace, figure):
+        layout = (2, 1)
+
+        figure.set_subplots(
+            dimensions=workspace.display_data.dimensions,
+            subplots=layout
+        )
+
+        figure.subplot_imshow(
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Worth checking, but I think that 2D single-channel images are not displayed as grayscale unless `colormap="gray"` is specified.

File: cellprofiler/modules/gammacorrection.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            dimensions=workspace.display_data.dimensions,
+            image=workspace.display_data.x_data,
+            x=0,
+            y=0
+        )
+
+        figure.subplot_imshow(
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Worth checking, but I think that 2D single-channel images are not displayed as grayscale unless `colormap="gray"` is specified.

File: tests/modules/test_gammacorrection.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import cellprofiler.modules.gammacorrection
+import numpy.testing
+import skimage.exposure
+
+instance = cellprofiler.modules.gammacorrection.GammaCorrection()
+
+
+def test_run(image, module, image_set, workspace):
+    module.x_name.value = "example"
```
Comment (mcquin) [inline tests/modules/test_gammacorrection.py]: I'm seeing this a lot. It's not really used in the tests. Does it make more sense in `conftest.py`?

File: cellprofiler/modules/gammacorrection.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+
+"""
+
+<strong>Gamma correction</strong>
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Module needs documentation.

File: cellprofiler/modules/gammacorrection.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        self.y_name = cellprofiler.setting.ImageNameProvider(
+            "Output",
+            "GammaCorrection"
+        )
+
+        self.gamma = cellprofiler.setting.Float(
+            "Gamma",
+            1,
+            minval=0.0,
+            maxval=100.0,
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Setting needs documentation.

File: cellprofiler/modules/gammacorrection.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        )
+
+        self.gain = cellprofiler.setting.Float(
```
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Setting needs documentation.

File: tests/modules/test_gammacorrection.py (hunk 6 above)
Comment (0x00b1) [inline tests/modules/test_gammacorrection.py]: Good idea!

File: cellprofiler/modules/gammacorrection.py (hunk 5 above)
Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: By design. It supports multi-channel.

File: cellprofiler/modules/gammacorrection.py (hunk 4 above)
Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: By design. It supports multi-channel.

File: cellprofiler/modules/gammacorrection.py (hunk 9 above)
Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: boo. :(

File: cellprofiler/modules/gammacorrection.py (hunk 8 above)
Comment (0x00b1) [inline cellprofiler/modules/gammacorrection.py]: boo. :(

File: cellprofiler/modules/gammacorrection.py (hunk 4 above)
Comment (mcquin) [inline cellprofiler/modules/gammacorrection.py]: Correct, and the "grayscale" images are displayed in... not black and white. Which might be unexpected to users.

File: cellprofiler/modules/gammacorrection.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import skimage.exposure
+
+
+class GammaCorrection(cellprofiler.module.ImageProcessing):
+    module_name = "GammaCorrection"
+
+    variable_revision_number = 1
+
+    def create_settings(self):
+        super(GammaCorrection, self).create_settings()
```
Comment (AnneCarpenter) [inline cellprofiler/modules/gammacorrection.py]: Should be `duper`

File: cellprofiler/module.py [hunk 11]
```diff
@@ -876,3 +877,77 @@ def on_deactivated(self):
...
+
+        x = images.get_image(x_name)
+
+        dimensions = x.dimensions
+
+        x_data = x.pixel_data
+
+        args = (setting.value for setting in self.settings()[2:])
+
+        y_data = self.function(x_data, *args)
```
Comment (mcquin) [inline cellprofiler/module.py]: Interesting! I like where this is going.

I think that `args` should be set from `self.visible_settings`, for the case where there are multiple operations per module. 

Additionally, how do we add arguments like `multichannel`, which are required but not determined from a setting? I am not sure how many of the strictly "Image Processing" modules use a `multichannel` option, so let's wait until we need to implement this.

**Comment by mcquin** [comment]
This module is missing documentation. 😢

**Pull request merged**